// db/db.js
const secrets = require('secrets')
const mongoose = require('mongoose')
const uuid = require('uuid/v4')
let functions = require('functions')
let Uri = functions['uri.js']
let uri = Uri(
	secrets.mongoDB, 
	{ 
		useMongoClient: true 
	}
)
const db = {
	mongoose, 
	connection: mongoose.createConnection(
		uri,
		{
			useNewUrlParser: true
		}
	),
	uri,
	schemas: [],
	models: {}
}

db.metaSchemaToMongoose = function(args){
	if(args.schema.schema){
		let schema = {}
		args.schema.schema.properties.forEach(property=>{
			schema[property.path] = property.mongoose.schema
		})
		if(schema){
			if(!db.models[args.schema.schema.path]){
				let s = new db.mongoose.Schema(schema)
				db.schemas.push(s)
				let m = db.connection.model(args.schema.schema.path, s)
				db.schemas[args.schema.schema.path] = s
				db.models[args.schema.schema.path] = m
			}
		}
	}
}
// model test
db.test = function(args){
		var test = new db.models['thing']({
			title: 'hey',
			price: 'nine ninety nine',
			thumbnail: 'lol.jpg'
		})

		test
		.save(err=>{
			if(err) {
				console.error(err)
			}
		})
	}
	// test()
module.exports = function(args){
	var args = args || {}
	if(args.schemas){
		args.schemas.forEach(schema=>{
			db.metaSchemaToMongoose({schema})
		})
	}
	return db
}
