// muppets/muppets.js

let puppeteer = require('puppeteer')
module.exports = async function(args){
	global.browser = await puppeteer.launch({headless: true})
	var args = args || {}
	let muppets = {
		"ready": [],
		muppets: [],
		min: args.muppets || 1,
		"running": {},
		init(args){
			this["find muppets"]({number: args.min || this.min})
			this.log(args)
			this.poolAI(args)
			return this
		},
		max: args.max || 4,
		poolAI(args){
			this.poolAI.loop = this.poolAI.loop || setInterval(()=>{
				if(this["ready"].length < this.min && this.muppets.length < this.max){
					this["find muppets"]({number: this.min - this["ready"].length})
				}
			}, 100)
		},
		available(args){
			var args = args || {}
			return new Promise((resolve, reject)=>{
				let i = setInterval(()=>{
					if(this["ready"].length > 0){
						clearInterval(i)
						let muppet = this["ready"].shift()
						let path = muppet.id
						if(!this["running"][path]){
							this["running"][path] = muppet
						}
						resolve(muppet)
					}
				}, 50)
			})
		},
		async "find muppets"(args){
			if(typeof args.number !== undefined && typeof args.number == 'number' && !this.finding){
				console.log(`[muppets][info] finding ${args.number} ${args.number == 1 ? 'muppet' : 'muppets'}`)
				this.finding = true
				let loop = new Array(args.number).fill(0)
				for await(var i of loop){
					let muppet = await {
						async available(args){
							var muppet = this
							await muppet.page.close()
							var toDeleteFrom = ['ready', 'muppets', 'running']
							for(var from of toDeleteFrom){
								var i = 0
								if(muppet.muppets[from] instanceof Array){
									for(var thing of muppet.muppets[from]){
										if(muppet.id == thing.id){
											muppet.muppets[from].splice(i, 1)
										}
										i++
									}
								} else {
									delete muppet.muppets[from][muppet.id]
								}
							}
						},
						muppets,
						id: Math.random(),
						getBrowser(){
							return global.browser
						},
						puppeteer
					}
					muppet.pageCrash = async function () {
						console.error("[muppets][error][puppeteer][page] Page crashed")
						try {
							console.log("[muppets][info][puppeteer][page] Trying soft page reload")
							await muppet.page.reload(); // soft fix
						} catch (recoveringErr) {
								console.error("[muppets][error][puppeteer][page] Soft page reload didn't work")
								// unable to reload the page, hard fix
								try {
									await muppet.browser.close();
								} catch (err) { // browser close was not necessary
										// you might want to log this error
								}
								console.log("[muppets][info][puppeteer][browser] Launching new browser")
								global.browser = muppet.browser = await muppet.puppeteer.launch({ headless: true });
						}							
					}
					muppets["ready"].push(muppet)
					muppets.muppets.push(muppet)
				}
				this.finding = false
			}
		},
		log(args){
			let prevmuppets = undefined
			this.log.loop = this.log.loop || setInterval(() => {
				if(prevmuppets !== this["ready"].length){
					console.log(`[muppets][info] ${this["ready"].length} ${this["ready"].length == 1 ? 'muppet' : 'muppets'} ready, ${Object.keys(this["running"]).length} running, ${this.muppets.length} in total.`)
					prevmuppets = this["ready"].length
				}
			}, 100);
			return `${this["ready"].length} ${this["ready"].length == 1 ? 'muppet' : 'muppets'} ready, ${Object.keys(this["running"]).length} running, ${this.muppets.length} in total.`
		}
	}
	

	return muppets.init(args)
}
