// schemabase/node_modules/marsHydroProductSchema.js
let productSchema = require('./product.js')
let functions = require('functions')
let smarts = require('smarts')()
let defaultParser = functions['defaultParser.js']

function priceParser(args){
	var ret = undefined
	if(args.val && typeof args.val == 'string'){
		let currency = args.val.indexOf('US') ? 'USD' : 'USD'
		ret = {
			name: args.property.name,
			values: {},
			source: 'mars hydro'
		}
		ret.values[currency] =  parseFloat(args.val.replace(/[\nUS\$]/ig, ''))
	}
	return ret
}
function typeParser(args1){
	return function(args){
		return args1.ret
	}
}

module.exports = function(){
	return {
		meta: {
			parsers: {
				typeParser
			}
		},
		"path" : "growlights/"+smarts.getsmart(global, 'env.level', 'dev')+"/products",
		schema: productSchema,
		"properties": [
			{
				path: 'title',
				html: {
					selector: ".product-shop > .f-fix > .product-primary > .product-name > a",
					attribute: 'innerHTML',
				}
			},
			{
				path: 'names',
			},
			{
				path: 'stock',
				html: {
					selector: "",
					attribute: '',
				}
			},
			{
				path: 'type',
				html: {
					page: true,
					selector: "body > div.wrapper > div > div.main-container.container.col2-right-layout > div > div.col-main > div.page-title.category-title > h1",
				}
			},
			{
				path: 'types',
				html: {
					properties: [
						{
							parser: typeParser({
								ret: 'product'
							})
						},
						{
							html: {
								page: true,
								selector: "body > div.wrapper > div > div.main-container.container.col2-right-layout > div > div.col-main > div.page-title.category-title > h1",
							}			
						}
					]
				}
			},
			{
				path: 'subtitle',
				html: {
					selector: "",
					attribute: '',
				}
			},
			{
				path: 'short description',
				html: {
					selector: ".product-shop > .f-fix > .desc",
					attribute: 'innerHTML',
				}
			},
			{
				path: 'prices',
				html: {
					properties: [
						{
							name: 'mars hydro old price',
							parser: priceParser,
							html: {
								selector: ".product-shop > .f-fix > .product-secondary > .price-box > .old-price > .price",
								attribute: 'innerHTML',
							}
						},
						{
							name: 'mars hydro regular price',
							parser: priceParser,
							html: {
								selector: ".product-shop > .f-fix > .product-secondary > .price-box > .regular-price > .price",
								attribute: 'innerHTML',
							}
						},
						{
							name: 'mars hydro sale price',
							parser: priceParser,
							html: {
								selector: ".product-shop > .f-fix > .product-secondary > .price-box > .special-price > .price",
								attribute: 'innerHTML',
							}
						},
					],
				}
			},
			{
				path: 'thumbnail',
				html: {
					selector: ".product-image > img",
					attribute: 'src',
				}
			},
			{
				path: 'images',
			},
			{
				path: 'source',
				html: {
					page: true,
					selector: "#header > div > div.row > div.col-md-10.col-xs-12.col-sm-12 > div.col-md-6.col-xs-12.col-sm-12 > a > img.large",
					attribute: 'alt',
				}
			}
		],
		methods: {
			pricesai: functions['productpricesai.js']
		},
		ai: functions['productai.js']
	}
}
