// sites/node_modules/quickbloomlights.js
let { quickBloomLightsProductSchema } = require('schemas')()
let s = require('smarts')()
function createUrl(args) {
	var ret = ''
	if(args.self){
		var args = args || {}
		var n = args.n || 0
		args.self.urls[n].forEach(part=>{
			ret += part
		})
	}
	return ret
}
let pages = {
	"californialightworks": {
		userAgent: "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.121 Safari/537.36",
		urls: [["https://www.quickbloomlights.com.au/","collections", "/", "californialightworks","?",""]],
		createUrl,
		list: {
			selector: '#shopify-section-collection__main > div > div.collection-matrix__wrapper > div'
		},
		"schema": s.mod(quickBloomLightsProductSchema(), (schema)=>{
			s.pushThings({
				options: [
					{
						parser: (val)=>{ return 'quick bloom lights california light works led growlights' }
					},
				],
				list: s.getsmart(
					s.getThing({
						option: {
							path: 'types'
						},
						list: s.getsmart(schema, 'properties', undefined),
						keys: ['path']
					}),
					'html.properties',
					undefined
				)
			})
			return schema

		})
	},
	"horticulture-lighting-group-led-grow-lights": {
		userAgent: "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.121 Safari/537.36",
		urls: [["https://www.quickbloomlights.com.au/","collections", "/", "horticulture-lighting-group-led-grow-lights","?",""]],
		createUrl,
		list: {
			selector: '#shopify-section-collection__main > div > div.collection-matrix__wrapper > div'
		},
		"schema": s.mod(quickBloomLightsProductSchema(), (schema)=>{
			s.pushThings({
				options: [
					{
						parser: (val)=>{ return 'quick bloom lights horticulture lighting group led growlights' }
					},
				],
				list: s.getsmart(
					s.getThing({
						option: {
							path: 'types'
						},
						list: s.getsmart(schema, 'properties', undefined),
						keys: ['path']
					}),
					'html.properties',
					undefined
				)
			})
			return schema

		})
	},
	"cunit-farm-led-grow-lights": {
		userAgent: "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.121 Safari/537.36",
		urls: [["https://www.quickbloomlights.com.au/","collections", "/", "unit-farm-led-grow-lights","?",""]],
		createUrl,
		list: {
			selector: '#shopify-section-collection__main > div > div.collection-matrix__wrapper > div'
		},
		"schema": s.mod(quickBloomLightsProductSchema(), (schema)=>{
			s.pushThings({
				options: [
					{
						parser: (val)=>{ return 'quick bloom lights cunit farm led growlights' }
					},
				],
				list: s.getsmart(
					s.getThing({
						option: {
							path: 'types'
						},
						list: s.getsmart(schema, 'properties', undefined),
						keys: ['path']
					}),
					'html.properties',
					undefined
				)
			})
			return schema

		})
	},
	"buy-optic-led-grow-lights-replace-hid-today": {
		userAgent: "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.121 Safari/537.36",
		urls: [["https://www.quickbloomlights.com.au/","collections", "/", "buy-optic-led-grow-lights-replace-hid-today","?",""]],
		createUrl,
		list: {
			selector: '#shopify-section-collection__main > div > div.collection-matrix__wrapper > div'
		},
		"schema": s.mod(quickBloomLightsProductSchema(), (schema)=>{
			s.pushThings({
				options: [
					{
						parser: (val)=>{ return 'quick bloom lights optic led growlights' }
					},
				],
				list: s.getsmart(
					s.getThing({
						option: {
							path: 'types'
						},
						list: s.getsmart(schema, 'properties', undefined),
						keys: ['path']
					}),
					'html.properties',
					undefined
				)
			})
			return schema

		})
	},
	"qbl-cob-led-grow-lights": {
		userAgent: "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.121 Safari/537.36",
		urls: [["https://www.quickbloomlights.com.au/","collections", "/", "qbl-cob-led-grow-lights","?",""]],
		createUrl,
		list: {
			selector: '#shopify-section-collection__main > div > div.collection-matrix__wrapper > div'
		},
		"schema": s.mod(quickBloomLightsProductSchema(), (schema)=>{
			s.pushThings({
				options: [
					{
						parser: (val)=>{ return 'quick bloom lights qbl cob led growlights' }
					},
				],
				list: s.getsmart(
					s.getThing({
						option: {
							path: 'types'
						},
						list: s.getsmart(schema, 'properties', undefined),
						keys: ['path']
					}),
					'html.properties',
					undefined
				)
			})
			return schema

		})
	},
	"black-dog-led-grow-lights": {
		userAgent: "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.121 Safari/537.36",
		urls: [["https://www.quickbloomlights.com.au/","collections", "/", "black-dog-led-grow-lights","?",""]],
		createUrl,
		list: {
			selector: '#shopify-section-collection__main > div > div.collection-matrix__wrapper > div'
		},
		"schema": s.mod(quickBloomLightsProductSchema(), (schema)=>{
			s.pushThings({
				options: [
					{
						parser: (val)=>{ return 'quick bloom lights black dog led growlights' }
					},
				],
				list: s.getsmart(
					s.getThing({
						option: {
							path: 'types'
						},
						list: s.getsmart(schema, 'properties', undefined),
						keys: ['path']
					}),
					'html.properties',
					undefined
				)
			})
			return schema

		})
	},
	"cob-led-grow-lights": {
		userAgent: "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.121 Safari/537.36",
		urls: [["https://www.quickbloomlights.com.au/","collections", "/", "cob-led-grow-lights","?",""]],
		createUrl,
		list: {
			selector: '#shopify-section-collection__main > div > div.collection-matrix__wrapper > div'
		},
		"schema": s.mod(quickBloomLightsProductSchema(), (schema)=>{
			s.pushThings({
				options: [
					{
						parser: (val)=>{ return 'quick bloom lights cob led growlights' }
					},
				],
				list: s.getsmart(
					s.getThing({
						option: {
							path: 'types'
						},
						list: s.getsmart(schema, 'properties', undefined),
						keys: ['path']
					}),
					'html.properties',
					undefined
				)
			})
			return schema

		})
	},
	"hydroponic-grow-tents": {
		userAgent: "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.121 Safari/537.36",
		urls: [["https://www.quickbloomlights.com.au/","collections", "/", "hydroponic-grow-tents","?",""]],
		createUrl,
		list: {
			selector: '#shopify-section-collection__main > div > div.collection-matrix__wrapper > div'
		},
		"schema": s.mod(quickBloomLightsProductSchema(), (schema)=>{
			s.pushThings({
				options: [
					{
						parser: (val)=>{ return 'quick bloom lights grow tents' }
					},
				],
				list: s.getsmart(
					s.getThing({
						option: {
							path: 'types'
						},
						list: s.getsmart(schema, 'properties', undefined),
						keys: ['path']
					}),
					'html.properties',
					undefined
				)
			})
			return schema

		})
	},
	"grow-tent-fans": {
		userAgent: "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.121 Safari/537.36",
		urls: [["https://www.quickbloomlights.com.au/","collections", "/", "grow-tent-fans","?",""]],
		createUrl,
		list: {
			selector: '#shopify-section-collection__main > div > div.collection-matrix__wrapper > div'
		},
		"schema": s.mod(quickBloomLightsProductSchema(), (schema)=>{
			s.pushThings({
				options: [
					{
						parser: (val)=>{ return 'fans' }
					},
					{
						parser: (val)=>{ return 'accessory' }
					},
					{
						parser: (val)=>{ return 'accessories' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow tent fans' }
					},
				],
				list: s.getsmart(
					s.getThing({
						option: {
							path: 'types'
						},
						list: s.getsmart(schema, 'properties', undefined),
						keys: ['path']
					}),
					'html.properties',
					undefined
				)
			})
			return schema

		})
	},
	"controllers": {
		userAgent: "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.121 Safari/537.36",
		urls: [["https://www.quickbloomlights.com.au/","collections", "/", "controllers","?",""]],
		createUrl,
		list: {
			selector: '#shopify-section-collection__main > div > div.collection-matrix__wrapper > div'
		},
		"schema": s.mod(quickBloomLightsProductSchema(), (schema)=>{
			s.pushThings({
				options: [
					{
						parser: (val)=>{ return 'quickbloomlights grow tent fan controllers' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow room fan controllers' }
					},
				],
				list: s.getsmart(
					s.getThing({
						option: {
							path: 'types'
						},
						list: s.getsmart(schema, 'properties', undefined),
						keys: ['path']
					}),
					'html.properties',
					undefined
				)
			})
			return schema

		})
	},
	"nutrients": {
		userAgent: "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.121 Safari/537.36",
		urls: [["https://www.quickbloomlights.com.au/","collections", "/", "nutrients","?",""]],
		createUrl,
		list: {
			selector: '#shopify-section-collection__main > div > div.collection-matrix__wrapper > div'
		},
		"schema": s.mod(quickBloomLightsProductSchema(), (schema)=>{
			s.pushThings({
				options: [
					{
						parser: (val)=>{ return 'quickbloomlights grow nutrients' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow additives' }
					},
				],
				list: s.getsmart(
					s.getThing({
						option: {
							path: 'types'
						},
						list: s.getsmart(schema, 'properties', undefined),
						keys: ['path']
					}),
					'html.properties',
					undefined
				)
			})
			return schema

		})
	},
	"additives-and-nutrient-supplements": {
		userAgent: "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.121 Safari/537.36",
		urls: [["https://www.quickbloomlights.com.au/","collections", "/", "additives-and-nutrient-supplements","?",""]],
		createUrl,
		list: {
			selector: '#shopify-section-collection__main > div > div.collection-matrix__wrapper > div'
		},
		"schema": s.mod(quickBloomLightsProductSchema(), (schema)=>{
			s.pushThings({
				options: [
					{
						parser: (val)=>{ return 'quickbloomlights grow nutrients' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow additives' }
					},
				],
				list: s.getsmart(
					s.getThing({
						option: {
							path: 'types'
						},
						list: s.getsmart(schema, 'properties', undefined),
						keys: ['path']
					}),
					'html.properties',
					undefined
				)
			})
			return schema

		})
	},
	"ph-ec-pens-and-tank-controllers": {
		userAgent: "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.121 Safari/537.36",
		urls: [["https://www.quickbloomlights.com.au/","collections", "/", "ph-ec-pens-and-tank-controllers","?",""]],
		createUrl,
		list: {
			selector: '#shopify-section-collection__main > div > div.collection-matrix__wrapper > div'
		},
		"schema": s.mod(quickBloomLightsProductSchema(), (schema)=>{
			s.pushThings({
				options: [
					{
						parser: (val)=>{ return 'grow accessory' }
					},
					{
						parser: (val)=>{ return 'grow accessories' }
					},
					{
						parser: (val)=>{ return 'PH/EC pens' }
					},
					{
						parser: (val)=>{ return 'PH EC pens' }
					},
					{
						parser: (val)=>{ return 'PH pens' }
					},
					{
						parser: (val)=>{ return 'PH monitors' }
					},
					{
						parser: (val)=>{ return 'PPM pens monitors' }
					},
					{
						parser: (val)=>{ return 'PH probes' }
					},
					{
						parser: (val)=>{ return 'PH meters' }
					},
					{
						parser: (val)=>{ return 'Nutrient meters' }
					},
					{
						parser: (val)=>{ return 'PH tests' }
					},
					{
						parser: (val)=>{ return 'EC pens' }
					},
				],
				list: s.getsmart(
					s.getThing({
						option: {
							path: 'types'
						},
						list: s.getsmart(schema, 'properties', undefined),
						keys: ['path']
					}),
					'html.properties',
					undefined
				)
			})
			return schema

		})
	}
	
}

let toExport = {
	name: "Quick Bloom Lights",
	pages,
}

module.exports = toExport
