// sites/node_modules/quickbloomlights.js
let { quickBloomLightsProductSchema } = require('schemabase')()
let s = require('smarts')()
let pages = {
	"californialightworks": {
		userAgent: "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.121 Safari/537.36",
		urls: [["https://www.quickbloomlights.com.au/","collections/californialightworks","?",""]],
		createUrl(args){
			var ret = ''
			if(args.self){
				var args = args || {}
				var n = args.n || 0
				args.self.urls[n].forEach(part=>{
					ret += part
				})
			}
			return ret
		},
		list: {
			selector: '#shopify-section-collection__main > div > div.collection-matrix__wrapper > div'
		},
		"schema": s.mod(quickBloomLightsProductSchema(), (schema)=>{
			s.pushThings({
				options: [
					{
						parser: (val)=>{ return 'led growlight' }
					},
					{
						parser: (val)=>{ return 'led growlights' }
					},
					{
						parser: (val)=>{ return 'growlight' }
					},
					{
						parser: (val)=>{ return 'growlights' }
					},
					{
						parser: (val)=>{ return 'led grow light' }
					},
					{
						parser: (val)=>{ return 'led grow lights' }
					},
					{
						parser: (val)=>{ return 'grow light' }
					},
					{
						parser: (val)=>{ return 'grow lights' }
					},
					{
						parser: (val)=>{ return 'led' }
					},
					{
						parser: (val)=>{ return 'leds' }
					},
					{
						parser: (val)=>{ return 'california light works led growlight' }
					},
					{
						parser: (val)=>{ return 'california light works led growlights' }
					},
					{
						parser: (val)=>{ return 'california light works growlight' }
					},
					{
						parser: (val)=>{ return 'california light works growlights' }
					},
					{
						parser: (val)=>{ return 'california light works led grow light' }
					},
					{
						parser: (val)=>{ return 'california light works led grow lights' }
					},
					{
						parser: (val)=>{ return 'california light works grow light' }
					},
					{
						parser: (val)=>{ return 'california light works grow lights' }
					},
					{
						parser: (val)=>{ return 'california light works led' }
					},
					{
						parser: (val)=>{ return 'california light works leds' }
					},
					{
						parser: (val)=>{ return 'california lightworks led growlight' }
					},
					{
						parser: (val)=>{ return 'california lightworks led growlights' }
					},
					{
						parser: (val)=>{ return 'california lightworks growlight' }
					},
					{
						parser: (val)=>{ return 'california lightworks growlights' }
					},
					{
						parser: (val)=>{ return 'california lightworks led grow light' }
					},
					{
						parser: (val)=>{ return 'california lightworks led grow lights' }
					},
					{
						parser: (val)=>{ return 'california lightworks grow light' }
					},
					{
						parser: (val)=>{ return 'california lightworks grow lights' }
					},
					{
						parser: (val)=>{ return 'california lightworks led' }
					},
					{
						parser: (val)=>{ return 'california lightworks leds' }
					},
				],
				list: s.getsmart(
					s.getThing({
						option: {
							path: 'types'
						},
						list: s.getsmart(schema, 'properties', undefined),
						keys: ['path']
					}),
					'html.properties',
					undefined
				)
			})
			return schema

		})
	},
	"horticulture-lighting-group-led-grow-lights": {
		userAgent: "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.121 Safari/537.36",
		urls: [["https://www.quickbloomlights.com.au/","collections/horticulture-lighting-group-led-grow-lights","?",""]],
		createUrl(args){
			var ret = ''
			if(args.self){
				var args = args || {}
				var n = args.n || 0
				args.self.urls[n].forEach(part=>{
					ret += part
				})
			}
			return ret
		},
		list: {
			selector: '#shopify-section-collection-template .blocklayout'
		},
		"schema": s.mod(quickBloomLightsProductSchema(), (schema)=>{
			s.pushThings({
				options: [
					{
						parser: (val)=>{ return 'led growlight' }
					},
					{
						parser: (val)=>{ return 'led growlights' }
					},
					{
						parser: (val)=>{ return 'growlight' }
					},
					{
						parser: (val)=>{ return 'growlights' }
					},
					{
						parser: (val)=>{ return 'led grow light' }
					},
					{
						parser: (val)=>{ return 'led grow lights' }
					},
					{
						parser: (val)=>{ return 'grow light' }
					},
					{
						parser: (val)=>{ return 'grow lights' }
					},
					{
						parser: (val)=>{ return 'led' }
					},
					{
						parser: (val)=>{ return 'leds' }
					},
					{
						parser: (val)=>{ return 'horticulture lighting group led growlight' }
					},
					{
						parser: (val)=>{ return 'horticulture lighting group led growlights' }
					},
					{
						parser: (val)=>{ return 'horticulture lighting group growlight' }
					},
					{
						parser: (val)=>{ return 'horticulture lighting group growlights' }
					},
					{
						parser: (val)=>{ return 'horticulture lighting group led grow light' }
					},
					{
						parser: (val)=>{ return 'horticulture lighting group led grow lights' }
					},
					{
						parser: (val)=>{ return 'horticulture lighting group grow light' }
					},
					{
						parser: (val)=>{ return 'horticulture lighting group grow lights' }
					},
					{
						parser: (val)=>{ return 'horticulture lighting group led' }
					},
					{
						parser: (val)=>{ return 'horticulture lighting group leds' }
					},
				],
				list: s.getsmart(
					s.getThing({
						option: {
							path: 'types'
						},
						list: s.getsmart(schema, 'properties', undefined),
						keys: ['path']
					}),
					'html.properties',
					undefined
				)
			})
			return schema

		})
	},
	"cunit-farm-led-grow-lights": {
		userAgent: "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.121 Safari/537.36",
		urls: [["https://www.quickbloomlights.com.au/","collections/unit-farm-led-grow-lights","?",""]],
		createUrl(args){
			var ret = ''
			if(args.self){
				var args = args || {}
				var n = args.n || 0
				args.self.urls[n].forEach(part=>{
					ret += part
				})
			}
			return ret
		},
		list: {
			selector: '#shopify-section-collection-template .blocklayout'
		},
		"schema": s.mod(quickBloomLightsProductSchema(), (schema)=>{
			s.pushThings({
				options: [
					{
						parser: (val)=>{ return 'led growlight' }
					},
					{
						parser: (val)=>{ return 'led growlights' }
					},
					{
						parser: (val)=>{ return 'growlight' }
					},
					{
						parser: (val)=>{ return 'growlights' }
					},
					{
						parser: (val)=>{ return 'led grow light' }
					},
					{
						parser: (val)=>{ return 'led grow lights' }
					},
					{
						parser: (val)=>{ return 'grow light' }
					},
					{
						parser: (val)=>{ return 'grow lights' }
					},
					{
						parser: (val)=>{ return 'led' }
					},
					{
						parser: (val)=>{ return 'leds' }
					},
					{
						parser: (val)=>{ return 'cunit farm led growlight' }
					},
					{
						parser: (val)=>{ return 'cunit farm led growlights' }
					},
					{
						parser: (val)=>{ return 'cunit farm growlight' }
					},
					{
						parser: (val)=>{ return 'cunit farm growlights' }
					},
					{
						parser: (val)=>{ return 'cunit farm led grow light' }
					},
					{
						parser: (val)=>{ return 'cunit farm led grow lights' }
					},
					{
						parser: (val)=>{ return 'cunit farm grow light' }
					},
					{
						parser: (val)=>{ return 'cunit farm grow lights' }
					},
					{
						parser: (val)=>{ return 'cunit farm led' }
					},
					{
						parser: (val)=>{ return 'cunit farm leds' }
					},
				],
				list: s.getsmart(
					s.getThing({
						option: {
							path: 'types'
						},
						list: s.getsmart(schema, 'properties', undefined),
						keys: ['path']
					}),
					'html.properties',
					undefined
				)
			})
			return schema

		})
	},
	"buy-optic-led-grow-lights-replace-hid-today": {
		userAgent: "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.121 Safari/537.36",
		urls: [["https://www.quickbloomlights.com.au/","collections/buy-optic-led-grow-lights-replace-hid-today","?",""]],
		createUrl(args){
			var ret = ''
			if(args.self){
				var args = args || {}
				var n = args.n || 0
				args.self.urls[n].forEach(part=>{
					ret += part
				})
			}
			return ret
		},
		list: {
			selector: '#shopify-section-collection-template .blocklayout'
		},
		"schema": s.mod(quickBloomLightsProductSchema(), (schema)=>{
			s.pushThings({
				options: [
					{
						parser: (val)=>{ return 'led growlight' }
					},
					{
						parser: (val)=>{ return 'led growlights' }
					},
					{
						parser: (val)=>{ return 'growlight' }
					},
					{
						parser: (val)=>{ return 'growlights' }
					},
					{
						parser: (val)=>{ return 'led grow light' }
					},
					{
						parser: (val)=>{ return 'led grow lights' }
					},
					{
						parser: (val)=>{ return 'grow light' }
					},
					{
						parser: (val)=>{ return 'grow lights' }
					},
					{
						parser: (val)=>{ return 'led' }
					},
					{
						parser: (val)=>{ return 'leds' }
					},
					{
						parser: (val)=>{ return 'optic led growlight' }
					},
					{
						parser: (val)=>{ return 'optic led growlights' }
					},
					{
						parser: (val)=>{ return 'optic growlight' }
					},
					{
						parser: (val)=>{ return 'optic growlights' }
					},
					{
						parser: (val)=>{ return 'optic led grow light' }
					},
					{
						parser: (val)=>{ return 'optic led grow lights' }
					},
					{
						parser: (val)=>{ return 'optic grow light' }
					},
					{
						parser: (val)=>{ return 'optic grow lights' }
					},
					{
						parser: (val)=>{ return 'optic led' }
					},
					{
						parser: (val)=>{ return 'optic leds' }
					},
				],
				list: s.getsmart(
					s.getThing({
						option: {
							path: 'types'
						},
						list: s.getsmart(schema, 'properties', undefined),
						keys: ['path']
					}),
					'html.properties',
					undefined
				)
			})
			return schema

		})
	},
	"qbl-cob-led-grow-lights": {
		userAgent: "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.121 Safari/537.36",
		urls: [["https://www.quickbloomlights.com.au/","collections/qbl-cob-led-grow-lights","?",""]],
		createUrl(args){
			var ret = ''
			if(args.self){
				var args = args || {}
				var n = args.n || 0
				args.self.urls[n].forEach(part=>{
					ret += part
				})
			}
			return ret
		},
		list: {
			selector: '#shopify-section-collection-template .blocklayout'
		},
		"schema": s.mod(quickBloomLightsProductSchema(), (schema)=>{
			s.pushThings({
				options: [
					{
						parser: (val)=>{ return 'led growlight' }
					},
					{
						parser: (val)=>{ return 'led growlights' }
					},
					{
						parser: (val)=>{ return 'growlight' }
					},
					{
						parser: (val)=>{ return 'growlights' }
					},
					{
						parser: (val)=>{ return 'led grow light' }
					},
					{
						parser: (val)=>{ return 'led grow lights' }
					},
					{
						parser: (val)=>{ return 'grow light' }
					},
					{
						parser: (val)=>{ return 'grow lights' }
					},
					{
						parser: (val)=>{ return 'led' }
					},
					{
						parser: (val)=>{ return 'leds' }
					},
					{
						parser: (val)=>{ return 'qbl cob led growlight' }
					},
					{
						parser: (val)=>{ return 'qbl cob led growlights' }
					},
					{
						parser: (val)=>{ return 'qbl cob growlight' }
					},
					{
						parser: (val)=>{ return 'qbl cob growlights' }
					},
					{
						parser: (val)=>{ return 'qbl cob led grow light' }
					},
					{
						parser: (val)=>{ return 'qbl cob led grow lights' }
					},
					{
						parser: (val)=>{ return 'qbl cob grow light' }
					},
					{
						parser: (val)=>{ return 'qbl cob grow lights' }
					},
					{
						parser: (val)=>{ return 'qbl cob led' }
					},
					{
						parser: (val)=>{ return 'qbl cob leds' }
					},
				],
				list: s.getsmart(
					s.getThing({
						option: {
							path: 'types'
						},
						list: s.getsmart(schema, 'properties', undefined),
						keys: ['path']
					}),
					'html.properties',
					undefined
				)
			})
			return schema

		})
	},
	"black-dog-led-grow-lights": {
		userAgent: "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.121 Safari/537.36",
		urls: [["https://www.quickbloomlights.com.au/","collections/black-dog-led-grow-lights","?",""]],
		createUrl(args){
			var ret = ''
			if(args.self){
				var args = args || {}
				var n = args.n || 0
				args.self.urls[n].forEach(part=>{
					ret += part
				})
			}
			return ret
		},
		list: {
			selector: '#shopify-section-collection-template .blocklayout'
		},
		"schema": s.mod(quickBloomLightsProductSchema(), (schema)=>{
			s.pushThings({
				options: [
					{
						parser: (val)=>{ return 'led growlight' }
					},
					{
						parser: (val)=>{ return 'led growlights' }
					},
					{
						parser: (val)=>{ return 'growlight' }
					},
					{
						parser: (val)=>{ return 'growlights' }
					},
					{
						parser: (val)=>{ return 'led grow light' }
					},
					{
						parser: (val)=>{ return 'led grow lights' }
					},
					{
						parser: (val)=>{ return 'grow light' }
					},
					{
						parser: (val)=>{ return 'grow lights' }
					},
					{
						parser: (val)=>{ return 'led' }
					},
					{
						parser: (val)=>{ return 'leds' }
					},
					{
						parser: (val)=>{ return 'black dog led growlight' }
					},
					{
						parser: (val)=>{ return 'black dog led growlights' }
					},
					{
						parser: (val)=>{ return 'black dog growlight' }
					},
					{
						parser: (val)=>{ return 'black dog growlights' }
					},
					{
						parser: (val)=>{ return 'black dog led grow light' }
					},
					{
						parser: (val)=>{ return 'black dog led grow lights' }
					},
					{
						parser: (val)=>{ return 'black dog grow light' }
					},
					{
						parser: (val)=>{ return 'black dog grow lights' }
					},
					{
						parser: (val)=>{ return 'black dog led' }
					},
					{
						parser: (val)=>{ return 'black dog leds' }
					},
				],
				list: s.getsmart(
					s.getThing({
						option: {
							path: 'types'
						},
						list: s.getsmart(schema, 'properties', undefined),
						keys: ['path']
					}),
					'html.properties',
					undefined
				)
			})
			return schema

		})
	},
	"cob-led-grow-lights": {
		userAgent: "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.121 Safari/537.36",
		urls: [["https://www.quickbloomlights.com.au/","collections/cob-led-grow-lights","?",""]],
		createUrl(args){
			var ret = ''
			if(args.self){
				var args = args || {}
				var n = args.n || 0
				args.self.urls[n].forEach(part=>{
					ret += part
				})
			}
			return ret
		},
		list: {
			selector: '#shopify-section-collection-template .blocklayout'
		},
		"schema": s.mod(quickBloomLightsProductSchema(), (schema)=>{
			s.pushThings({
				options: [
					{
						parser: (val)=>{ return 'led growlight' }
					},
					{
						parser: (val)=>{ return 'led growlights' }
					},
					{
						parser: (val)=>{ return 'growlight' }
					},
					{
						parser: (val)=>{ return 'growlights' }
					},
					{
						parser: (val)=>{ return 'led grow light' }
					},
					{
						parser: (val)=>{ return 'led grow lights' }
					},
					{
						parser: (val)=>{ return 'grow light' }
					},
					{
						parser: (val)=>{ return 'grow lights' }
					},
					{
						parser: (val)=>{ return 'led' }
					},
					{
						parser: (val)=>{ return 'leds' }
					},
					{
						parser: (val)=>{ return 'cob led growlight' }
					},
					{
						parser: (val)=>{ return 'cob led growlights' }
					},
					{
						parser: (val)=>{ return 'cob growlight' }
					},
					{
						parser: (val)=>{ return 'cob growlights' }
					},
					{
						parser: (val)=>{ return 'cob led grow light' }
					},
					{
						parser: (val)=>{ return 'cob led grow lights' }
					},
					{
						parser: (val)=>{ return 'cob grow light' }
					},
					{
						parser: (val)=>{ return 'cob grow lights' }
					},
					{
						parser: (val)=>{ return 'cob led' }
					},
					{
						parser: (val)=>{ return 'cob leds' }
					},
				],
				list: s.getsmart(
					s.getThing({
						option: {
							path: 'types'
						},
						list: s.getsmart(schema, 'properties', undefined),
						keys: ['path']
					}),
					'html.properties',
					undefined
				)
			})
			return schema

		})
	},
	"hydroponic-grow-tents": {
		userAgent: "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.121 Safari/537.36",
		urls: [["https://www.quickbloomlights.com.au/","collections/hydroponic-grow-tents","?",""]],
		createUrl(args){
			var ret = ''
			if(args.self){
				var args = args || {}
				var n = args.n || 0
				args.self.urls[n].forEach(part=>{
					ret += part
				})
			}
			return ret
		},
		list: {
			selector: '#shopify-section-collection-template .blocklayout'
		},
		"schema": s.mod(quickBloomLightsProductSchema(), (schema)=>{
			s.pushThings({
				options: [
					{
						parser: (val)=>{ return 'grow tent' }
					},
					{
						parser: (val)=>{ return 'grow tents' }
					},
					{
						parser: (val)=>{ return 'tent' }
					},
					{
						parser: (val)=>{ return 'tents' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow tent' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow tents' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow tent' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow tents' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights tent' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights tents' }
					},
					{
						parser: (val)=>{ return 'grow room' }
					},
					{
						parser: (val)=>{ return 'grow rooms' }
					},
					{
						parser: (val)=>{ return 'room' }
					},
					{
						parser: (val)=>{ return 'rooms' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow room' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow rooms' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow room' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow rooms' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights room' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights tents' }
					},
				],
				list: s.getsmart(
					s.getThing({
						option: {
							path: 'types'
						},
						list: s.getsmart(schema, 'properties', undefined),
						keys: ['path']
					}),
					'html.properties',
					undefined
				)
			})
			return schema

		})
	},
	"grow-tent-fans": {
		userAgent: "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.121 Safari/537.36",
		urls: [["https://www.quickbloomlights.com.au/","collections/grow-tent-fans","?",""]],
		createUrl(args){
			var ret = ''
			if(args.self){
				var args = args || {}
				var n = args.n || 0
				args.self.urls[n].forEach(part=>{
					ret += part
				})
			}
			return ret
		},
		list: {
			selector: '#shopify-section-collection-template .blocklayout'
		},
		"schema": s.mod(quickBloomLightsProductSchema(), (schema)=>{
			s.pushThings({
				options: [
					{
						parser: (val)=>{ return 'grow tent fan' }
					},
					{
						parser: (val)=>{ return 'grow tent fans' }
					},
					{
						parser: (val)=>{ return 'grow tent fan' }
					},
					{
						parser: (val)=>{ return 'grow tent fans' }
					},
					{
						parser: (val)=>{ return 'tent fan' }
					},
					{
						parser: (val)=>{ return 'tent fans' }
					},
					{
						parser: (val)=>{ return 'fan' }
					},
					{
						parser: (val)=>{ return 'fans' }
					},
					{
						parser: (val)=>{ return 'accessory' }
					},
					{
						parser: (val)=>{ return 'accessories' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow tent fan' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow tent fans' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow tent fan' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow tent fans' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights tent fan' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights tent fans' }
					},
					{
						parser: (val)=>{ return 'grow room fan' }
					},
					{
						parser: (val)=>{ return 'grow room fans' }
					},
					{
						parser: (val)=>{ return 'grow room fan' }
					},
					{
						parser: (val)=>{ return 'grow room fans' }
					},
					{
						parser: (val)=>{ return 'room fan' }
					},
					{
						parser: (val)=>{ return 'room fans' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow room fan' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow room fans' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow room fan' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow room fans' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights room fan' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights tent fans' }
					},
				],
				list: s.getsmart(
					s.getThing({
						option: {
							path: 'types'
						},
						list: s.getsmart(schema, 'properties', undefined),
						keys: ['path']
					}),
					'html.properties',
					undefined
				)
			})
			return schema

		})
	},
	"controllers": {
		userAgent: "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.121 Safari/537.36",
		urls: [["https://www.quickbloomlights.com.au/","collections/controllers","?",""]],
		createUrl(args){
			var ret = ''
			if(args.self){
				var args = args || {}
				var n = args.n || 0
				args.self.urls[n].forEach(part=>{
					ret += part
				})
			}
			return ret
		},
		list: {
			selector: '#shopify-section-collection-template .blocklayout'
		},
		"schema": s.mod(quickBloomLightsProductSchema(), (schema)=>{
			s.pushThings({
				options: [
					{
						parser: (val)=>{ return 'grow tent fan controller' }
					},
					{
						parser: (val)=>{ return 'grow tent fan controllers' }
					},
					{
						parser: (val)=>{ return 'grow tent fan controller' }
					},
					{
						parser: (val)=>{ return 'grow tent fan controllers' }
					},
					{
						parser: (val)=>{ return 'tent fan controller' }
					},
					{
						parser: (val)=>{ return 'tent fan controllers' }
					},
					{
						parser: (val)=>{ return 'fan controller' }
					},
					{
						parser: (val)=>{ return 'fan controllers' }
					},
					{
						parser: (val)=>{ return 'accessory' }
					},
					{
						parser: (val)=>{ return 'accessories' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow tent fan controller' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow tent fan controllers' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow tent fan controller' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow tent fan controllers' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights tent fan controller' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights tent fan controllers' }
					},
					{
						parser: (val)=>{ return 'grow room fan controller' }
					},
					{
						parser: (val)=>{ return 'grow room fan controllers' }
					},
					{
						parser: (val)=>{ return 'grow room fan controller' }
					},
					{
						parser: (val)=>{ return 'grow room fan controllers' }
					},
					{
						parser: (val)=>{ return 'room fan controller' }
					},
					{
						parser: (val)=>{ return 'room fan controllers' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow room fan controller' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow room fan controllers' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow room fan controller' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow room fan controllers' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights room fan controller' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights tent fan controllers' }
					},
				],
				list: s.getsmart(
					s.getThing({
						option: {
							path: 'types'
						},
						list: s.getsmart(schema, 'properties', undefined),
						keys: ['path']
					}),
					'html.properties',
					undefined
				)
			})
			return schema

		})
	},
	"nutrients": {
		userAgent: "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.121 Safari/537.36",
		urls: [["https://www.quickbloomlights.com.au/","collections/nutrients","?",""]],
		createUrl(args){
			var ret = ''
			if(args.self){
				var args = args || {}
				var n = args.n || 0
				args.self.urls[n].forEach(part=>{
					ret += part
				})
			}
			return ret
		},
		list: {
			selector: '#shopify-section-collection-template .blocklayout'
		},
		"schema": s.mod(quickBloomLightsProductSchema(), (schema)=>{
			s.pushThings({
				options: [
					{
						parser: (val)=>{ return 'grow nutrients' }
					},
					{
						parser: (val)=>{ return 'grow additives' }
					},
					{
						parser: (val)=>{ return 'grow nutrient' }
					},
					{
						parser: (val)=>{ return 'grow additive' }
					},
					{
						parser: (val)=>{ return 'grow nutrients' }
					},
					{
						parser: (val)=>{ return 'grow additives' }
					},
					{
						parser: (val)=>{ return 'grow nutrient' }
					},
					{
						parser: (val)=>{ return 'grow additive' }
					},
					{
						parser: (val)=>{ return 'nutrients' }
					},
					{
						parser: (val)=>{ return 'additives' }
					},
					{
						parser: (val)=>{ return 'nutrient' }
					},
					{
						parser: (val)=>{ return 'additive' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow nutrients' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow additives' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow nutrient' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow additive' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow nutrients' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow additives' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow nutrient' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow additive' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights nutrients' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights additives' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights nutrient' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights additive' }
					},
				],
				list: s.getsmart(
					s.getThing({
						option: {
							path: 'types'
						},
						list: s.getsmart(schema, 'properties', undefined),
						keys: ['path']
					}),
					'html.properties',
					undefined
				)
			})
			return schema

		})
	},
	"additives-and-nutrient-supplements": {
		userAgent: "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.121 Safari/537.36",
		urls: [["https://www.quickbloomlights.com.au/","collections/additives-and-nutrient-supplements","?",""]],
		createUrl(args){
			var ret = ''
			if(args.self){
				var args = args || {}
				var n = args.n || 0
				args.self.urls[n].forEach(part=>{
					ret += part
				})
			}
			return ret
		},
		list: {
			selector: '#shopify-section-collection-template .blocklayout'
		},
		"schema": s.mod(quickBloomLightsProductSchema(), (schema)=>{
			s.pushThings({
				options: [
					{
						parser: (val)=>{ return 'grow nutrients' }
					},
					{
						parser: (val)=>{ return 'grow additives' }
					},
					{
						parser: (val)=>{ return 'grow nutrient' }
					},
					{
						parser: (val)=>{ return 'grow additive' }
					},
					{
						parser: (val)=>{ return 'grow nutrients' }
					},
					{
						parser: (val)=>{ return 'grow additives' }
					},
					{
						parser: (val)=>{ return 'grow nutrient' }
					},
					{
						parser: (val)=>{ return 'grow additive' }
					},
					{
						parser: (val)=>{ return 'nutrients' }
					},
					{
						parser: (val)=>{ return 'additives' }
					},
					{
						parser: (val)=>{ return 'nutrient' }
					},
					{
						parser: (val)=>{ return 'additive' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow nutrients' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow additives' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow nutrient' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow additive' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow nutrients' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow additives' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow nutrient' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights grow additive' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights nutrients' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights additives' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights nutrient' }
					},
					{
						parser: (val)=>{ return 'quickbloomlights additive' }
					},
				],
				list: s.getsmart(
					s.getThing({
						option: {
							path: 'types'
						},
						list: s.getsmart(schema, 'properties', undefined),
						keys: ['path']
					}),
					'html.properties',
					undefined
				)
			})
			return schema

		})
	},
	"ph-ec-pens-and-tank-controllers": {
		userAgent: "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.121 Safari/537.36",
		urls: [["https://www.quickbloomlights.com.au/","collections/ph-ec-pens-and-tank-controllers","?",""]],
		createUrl(args){
			var ret = ''
			if(args.self){
				var args = args || {}
				var n = args.n || 0
				args.self.urls[n].forEach(part=>{
					ret += part
				})
			}
			return ret
		},
		list: {
			selector: '#shopify-section-collection-template .blocklayout'
		},
		"schema": s.mod(quickBloomLightsProductSchema(), (schema)=>{
			s.pushThings({
				options: [
					{
						parser: (val)=>{ return 'grow accessory' }
					},
					{
						parser: (val)=>{ return 'grow accessories' }
					},
					{
						parser: (val)=>{ return 'grow accessory' }
					},
					{
						parser: (val)=>{ return 'grow accessories' }
					},
					{
						parser: (val)=>{ return 'PH/EC pen' }
					},
					{
						parser: (val)=>{ return 'PH/EC pens' }
					},
					{
						parser: (val)=>{ return 'PH EC pen' }
					},
					{
						parser: (val)=>{ return 'PH EC pens' }
					},
					{
						parser: (val)=>{ return 'PH pen' }
					},
					{
						parser: (val)=>{ return 'PH pens' }
					},
					{
						parser: (val)=>{ return 'PH monitor' }
					},
					{
						parser: (val)=>{ return 'PH monitors' }
					},
					{
						parser: (val)=>{ return 'monitor' }
					},
					{
						parser: (val)=>{ return 'monitors' }
					},
					{
						parser: (val)=>{ return 'PPM pen' }
					},
					{
						parser: (val)=>{ return 'PPM pens' }
					},
					{
						parser: (val)=>{ return 'PPM monitor' }
					},
					{
						parser: (val)=>{ return 'PPM monitors' }
					},
					{
						parser: (val)=>{ return 'PH probe' }
					},
					{
						parser: (val)=>{ return 'PH probes' }
					},
					{
						parser: (val)=>{ return 'PH meter' }
					},
					{
						parser: (val)=>{ return 'PH meters' }
					},
					{
						parser: (val)=>{ return 'Nutrient meter' }
					},
					{
						parser: (val)=>{ return 'Nutrient meters' }
					},
					{
						parser: (val)=>{ return 'PH test' }
					},
					{
						parser: (val)=>{ return 'PH tests' }
					},
					{
						parser: (val)=>{ return 'EC pen' }
					},
					{
						parser: (val)=>{ return 'EC pens' }
					},
				],
				list: s.getsmart(
					s.getThing({
						option: {
							path: 'types'
						},
						list: s.getsmart(schema, 'properties', undefined),
						keys: ['path']
					}),
					'html.properties',
					undefined
				)
			})
			return schema

		})
	}
	
}

let toExport = {
	name: "Quick Bloom Lights",
	pages,
}

module.exports = toExport
