// sites/node_modules/testsiteignore.js
let { marsHydroProductSchema } = require('schemabase')()
let s = require('smarts')()
let pages = {
	"led growlights": {
		userAgent: "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.121 Safari/537.36",
		urls: [["https://www.mars-hydro.com/","led-grow-light","?","limit=25"]],
		createUrl(args){
			var ret = ''
			if(args.self){
				var args = args || {}
				var n = args.n || 0
				args.self.urls[n].forEach(part=>{
					ret += part
				})
			}
			return ret
		},
		list: {
			selector: '#products-list'
		},
		"schema": s.mod(marsHydroProductSchema(), (schema)=>{
			s.pushThings({
				options: [
					{
						parser: (val)=>{ return 'led growlight' }
					},
				],
				list: s.getsmart(
					s.getThing({
						option: {
							path: 'types'
						},
						list: s.getsmart(schema, 'properties', undefined),
						keys: ['path']
					}),
					'html.properties',
					undefined
				)
			})
			return schema
		})
	},
}

let toExport =  {
	name: "Test Site",
	pages
}

module.exports = toExport
