// puppets/puppets.js

let puppeteer = require('puppeteer')

module.exports = async function(args = {}){
	global.logger.log("[greenroom][puppets] Getting puppets ready")
	global.logger.log("[greenroom][puppets][browser] Launching browser")
	let browser = await puppeteer.launch({
		headless: true,
		args: [
			"--unlimited-storage",
      "--full-memory-crash-report",
      "--disable-gpu",
      "--ignore-certificate-errors",
      "--no-sandbox",
      "--disable-setuid-sandbox",
      "--disable-dev-shm-usage",
      "--lang=en-US;q=0.9,en;q=0.8",
      "--user-agent=Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/68.0.3440.106 Safari/537.36",
		]
	})
	let puppets = {
		"ready": [],
		puppets: [],
		min: args.puppets || 1,
		"running": {},
		browser,
		async close(){
			global.logger.log("[greenroom][puppets][browser] Closing browser")
			await browser.close()
		},
		init(args){
			this["find puppets"]({number: args.min || this.min})
			this.log(args)
			this.poolAI(args)
			return this
		},
		max: args.max || 4,
		poolAI(args){
			this.poolAI.loop = this.poolAI.loop || setInterval(()=>{
				if(this["ready"].length < this.min && this.puppets.length < this.max){
					this["find puppets"]({number: this.min - this["ready"].length})
				}
			}, 100)
		},
		available(args){
			var args = args || {}
			return new Promise((resolve, reject)=>{
				let i = setInterval(()=>{
					if(this["ready"].length > 0){
						clearInterval(i)
						let puppet = this["ready"].shift()
						let path = puppet.id
						if(!this["running"][path]){
							this["running"][path] = puppet
						}
						resolve(puppet)
					}
				}, 50)
			})
		},
		"find puppets": async function (args){
			if(typeof args.number !== undefined && typeof args.number == 'number' && !this.finding){
				global.logger.log(`[greenroom][info] finding ${args.number} ${args.number == 1 ? 'puppet' : 'puppets'}`)
				this.finding = true
				let loop = new Array(args.number).fill(0)
				for await(var i of loop){
					let puppet = {
						id: Math.random(),
						puppets,
						puppeteer,
						getBrowser(){
							return browser
						},
						available: async function (args){
							var puppet = this
							await puppet.page.close()
							var toDeleteFrom = ['ready', 'puppets', 'running']
							for(var from of toDeleteFrom){
								var i = 0
								if(puppet.puppets[from] instanceof Array){
									for(var thing of puppet.puppets[from]){
										if(puppet.id == thing.id){
											puppet.puppets[from].splice(i, 1)
										}
										i++
									}
								} else {
									delete puppet.puppets[from][puppet.id]
								}
							}
						},
						pageCrash: async function () {
							global.logger.error("[greenroom][puppeteer][page][error] Page crashed")
							try {
								global.logger.log("[greenroom][puppeteer][page][info] Trying soft page reload")
								await puppet.page.reload(); // soft fix
							} catch (recoveringErr) {
									global.logger.error("[greenroom][puppeteer][page][error] Soft page reload didn't work")
									// unable to reload the page, hard fix
									try {
										await puppet.browser.close();
									} catch (err) { // browser close was not necessary
											// you might want to log this error
									}
									global.logger.log("[greenroom][puppeteer][browser][info] Launching new browser")
									browser = await puppet.puppeteer.launch({ headless: true });
							}							
						}
					}
					puppets["ready"].push(puppet)
					puppets.puppets.push(puppet)
				}
				this.finding = false
			}
		},
		log(args){
			let prevpuppets = undefined
			this.log.loop = this.log.loop || setInterval(() => {
				if(prevpuppets !== this["ready"].length){
					global.logger.log(`[greenroom][info] ${this["ready"].length} ${this["ready"].length == 1 ? 'puppet' : 'puppets'} ready, ${Object.keys(this["running"]).length} running, ${this.puppets.length} in total.`)
					prevpuppets = this["ready"].length
				}
			}, 100);
			return `${this["ready"].length} ${this["ready"].length == 1 ? 'puppet' : 'puppets'} ready, ${Object.keys(this["running"]).length} running, ${this.puppets.length} in total.`
		}
	}
	

	global.logger.log("[greenroom][puppets][info] Initialising puppets")
	puppets.init(args)

	global.logger.log("[greenroom][puppets][info] Puppets are ready")
	return puppets
	
}
